cartographer/mapping/id_test.cc

```diff
@@ -254,6 +254,53 @@ TEST(IdTest, LowerBoundFuzz) {
   }
 }
 
+TEST(IdTest, LowerBoundTrimmedTrajectory) {
+  constexpr int kTrajectoryId = 1;
+
+  std::mt19937 rng;
+  std::uniform_int_distribution<int> dt_dist(1, 20);
+
+  const int N = 500;
+  int t = 0;
+  MapById<SubmapId, Data> map_by_id;
+  for (int j = 0; j < N; ++j) {
+    t = t + dt_dist(rng);
+    map_by_id.Append(kTrajectoryId, Data(t));
+  }
+
+  // Choose random length of a trim segment.
+  std::uniform_int_distribution<int> dt_trim_segment_length(
+      1, static_cast<int>(N / 2));
+  size_t trim_segment_length = dt_trim_segment_length(rng);
+  // Choose random start for a trim_segment.
+  std::uniform_int_distribution<int> dt_trim_segment_start(
+      2, N - trim_segment_length - 1);
+  size_t trim_segment_start_index = dt_trim_segment_start(rng);
+
+  auto trim_segment_start = map_by_id.begin();
+  std::advance(trim_segment_start, trim_segment_start_index);
+
+  auto trim_segment_end = map_by_id.begin();
+  std::advance(trim_segment_end,
+               trim_segment_start_index + trim_segment_length);
+
+  for (auto it = trim_segment_start; it != trim_segment_end;) {
+    const auto this_it = it;
```

Comment:
This is necessary since after trimming, `it` no longer is a valid iterator, so `++it` after trimming cannot be done and we have to do it before.